bodhi/server/views/generic.py

```diff
@@ -68,6 +68,71 @@ def get_latest_updates(request, critpath, security):
     return query.limit(5).all()
 
 
+def get_update_counts(request, releaseid):
+    db = request.db
+    query = db.query(models.Update)
+
+    release = models.Release.get(releaseid, request.db)
+
+    basequery = query.filter(
+        models.Update.release == release)
+
+    pending_basequery = basequery.filter(
+        models.Update.status == models.UpdateStatus.pending)
+    pending_updates_total = pending_basequery.count()
+    pending_newpackage_total = pending_basequery\
+        .filter(models.Update.type == models.UpdateType.newpackage).count()
+    pending_bugfix_total = pending_basequery\
+        .filter(models.Update.type == models.UpdateType.bugfix).count()
+    pending_enhancement_total = pending_basequery\
+        .filter(models.Update.type == models.UpdateType.enhancement).count()
+    pending_security_total = pending_basequery\
+        .filter(models.Update.type == models.UpdateType.security).count()
+
+    testing_basequery = basequery.filter(
+        models.Update.status == models.UpdateStatus.testing)
+    testing_updates_total = testing_basequery.count()
+    testing_newpackage_total = testing_basequery\
+        .filter(models.Update.type == models.UpdateType.newpackage).count()
+    testing_bugfix_total = testing_basequery\
+        .filter(models.Update.type == models.UpdateType.bugfix).count()
+    testing_enhancement_total = testing_basequery\
+        .filter(models.Update.type == models.UpdateType.enhancement).count()
+    testing_security_total = testing_basequery\
+        .filter(models.Update.type == models.UpdateType.security).count()
+
+    stable_basequery = basequery.filter(
+        models.Update.status == models.UpdateStatus.stable)
+    stable_updates_total = stable_basequery.count()
+    stable_newpackage_total = stable_basequery\
+        .filter(models.Update.type == models.UpdateType.newpackage).count()
+    stable_bugfix_total = stable_basequery\
+        .filter(models.Update.type == models.UpdateType.bugfix).count()
+    stable_enhancement_total = stable_basequery\
+        .filter(models.Update.type == models.UpdateType.enhancement).count()
+    stable_security_total = stable_basequery\
+        .filter(models.Update.type == models.UpdateType.security).count()
```

Comment:
This function is a bit long and a bit repetitive. We should make a helper function to do the heavy lifting and then call it three times, once for each block above. Something like this:

```
def _get_status_counts(basequery, status):
    """
    Return a dictionary with the counts of objects found in the basequery, specified by total count, newpackage count, bugfix count, enhancement count, and security count. The dictionary keys will be named with the template {status}_{type}_total. For example, if status is models.UpdateStatus.stable, a dictionary with the following keys would be returned: stable_updates_total, stable_newpackage_total, stable_bugfix_total, stable_enhancement_total, stable_security_total.

    Args:
        basequery (uh, sqlalchemy.something.something? Use type(basequery) to find out what it is…): The basequery of updates we want to count and further filter on.
        status (bodhi.server.models.UpdateStatus): The update status we want to filter by in basequery
    Return:
        dict: A dictionary describing the counts of the updates, as described above.
    """
    basequery = basequery.filter(models.Update.status == status)
    return {
        '{}_updates_total'.format(status): basequery.count()
        '{}_newpackage_total'.format(status): basequery.filter(models.Update.type == models.UpdateType.newpackage).count()
        … and so forth
    }
```

Then you could just make three calls to it, and keep all the code shorter and simpler. Of course, you'd need to aggregate the return results into one dictionary, but you can use the [```dict.update```](https://docs.python.org/2/library/stdtypes.html#dict.update) method for that to make it nice and clean.